Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -109,3 +113,79 @@ theorem eq_one_or_neg_one_of_unit_of_congruent (hcong : ∃ n : ℤ, λ ^ 2 ∣
     have : (hζ.pow_of_coprime 2 (by decide)).toInteger = hζ.toInteger ^ 2 := by ext; simp
     simp only [this, PNat.val_ofNat, Nat.cast_ofNat, mul_neg, Int.cast_neg, ← neg_add, ←
       sub_eq_iff_eq_add.1 hx, Units.val_neg, val_pow_eq_pow_val, IsUnit.unit_spec, neg_neg]
+
+variable (x : 𝓞 K)
+
+/-- Let `(x : 𝓞 K)`. Then we have that `λ` divides one amongst `x`, `x - 1` and `x + 1`. -/
+lemma dvd_or_dvd_sub_one_or_dvd_add_one : λ ∣ x ∨ λ ∣ x - 1 ∨ λ ∣ x + 1 := by
+  classical
+  let _ := hζ.fintypeQuotienttoIntegerSubOne (by decide)
+  have := Finset.mem_univ (Ideal.Quotient.mk (Ideal.span {λ}) x)
+  rw [Finset.univ_of_card_eq_three] at this
+  · simp only [Finset.mem_insert, Finset.mem_singleton] at this
+    rcases this with (h | h | h)
+    · left
+      exact Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 h
+    · right; left
+      refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+      rw [RingHom.map_sub, h, RingHom.map_one, sub_self]
+    · right; right
+      refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+      rw [RingHom.map_add, h, RingHom.map_one, add_left_neg]
+  · rw [hζ.card_quotient_toInteger_sub_one, hζ.norm_toInteger_sub_one_of_prime_ne_two' (by decide)]
+    simp
+
+/-- We have that `η ^ 2 + η + 1 = 0`. -/
+lemma eta_eq_add_eta_add_one : (η : 𝓞 K) ^ 2 + η + 1 = 0 := by
+  ext; simpa using hζ.isRoot_cyclotomic (by decide)
+
+/-- We have that `x ^ 3 - 1 = (x - 1) * (x - η) * (x - η ^ 2)`. -/
+lemma cube_sub_one_eq_mul : x ^ 3 - 1 = (x - 1) * (x - η) * (x - η ^ 2) := by
+  symm
+  calc _ = x ^ 3 - x ^ 2 * (η ^ 2 + η + 1) + x * (η ^ 2 + η + η ^ 3) - η ^ 3 := by ring
+  _ = x ^ 3 - x ^ 2 * (η ^ 2 + η + 1) + x * (η ^ 2 + η + 1) - 1 := by
+    simp [show hζ.toInteger ^ 3 = 1 from hζ.toInteger_isPrimitiveRoot.pow_eq_one]
```

Comment:
Because `hζ.toInteger_isPrimitiveRoot` says that `hζ.toInteger ^ ↑3 = 1`, where `3 : ℕ+` but in the goal there is a `3 : ℕ`. The general statement of `toInteger_isPrimitiveRoot` is I think correct (positivity is needed in the definition of `toInteger`).